### PR TITLE
Synchronize distributed token sampling across model-parallel ranks

### LINKED ISSUE #270: Distributed Torch and Triton sampling can diverge across model-parallel ranks
## Summary

The reference Torch and Triton token generators perform stochastic next-token sampling independently on every distributed rank.

Both model-parallel implementations cooperate across ranks during the MLP and therefore require every rank to continue with the same token sequence. With `temperature > 0`, each rank currently calls `torch.multinomial()` locally and no sampled token is broadcast.

`gpt_oss.chat` invokes `generator.generate(...)` without overriding temperature, so it uses the generators' default `temperature=1.0`. A multi-GPU chat run can therefore choose different next tokens on different ranks after the first sampling step.

## Current behavior

Both generators contain the equivalent of:

```python
probs = torch.softmax(logits * (1.0 / temperature), dim=-1)
predicted_token = torch.multinomial(probs, num_samples=1).item()
```

There is no distributed synchronization of `predicted_token` before it becomes the next input token.

## Impact

If model-parallel ranks sample different tokens, subsequent forward passes combine shard computations for different token histories. The distributed inference result is no longer a coherent model sequence and can become nondeterministic/corrupted under the default interactive-chat sampling path.

Greedy `temperature=0` generation happens to avoid the divergence because identical logits produce the same argmax on each rank.

## Proposed resolution

When `torch.distributed` is initialized, sample/select the next token on rank 0 and broadcast that token to every rank before continuing generation. Keep the existing local sampling path when distributed inference is not initialized.

Apply the same synchronization policy to both the Torch and Triton reference generators and add CPU-only mocked regressions showing nonzero ranks use the token supplied by rank 0 rather than sampling independently.

### PR DESCRIPTION
**Pull request by sylvesterkaczmarek**
## Summary

Keep stochastic generation on the same token history across model-parallel ranks.

The Torch and Triton reference generators currently call `torch.multinomial()` independently on every distributed rank. With nonzero temperature, ranks can sample different tokens and then feed different token histories into later model-parallel collectives.

Fixes #270.

## Fix

Add one shared sampling helper used by both reference backends:

- in single-process mode, preserve the existing greedy and stochastic sampling behavior;
- when `torch.distributed` is initialized, sample/select the next token only on rank 0;
- broadcast that token to every rank before continuing generation.

Both the Torch and Triton generators now use the same synchronized path.

## Regression coverage

Adds CPU-only mocked tests verifying that:

- local greedy sampling is unchanged;
- rank 0 samples and broadcasts its token;
- a nonzero rank does not call `torch.multinomial()` and instead consumes the token received from rank 0.

Model logits, temperature scaling, stop-token handling, log-probability reporting, and single-process generation semantics are otherwise unchanged.

### CODE REVIEW

**Comment by sylvesterkaczmarek** [comment]
Superseded by #291. Codex correctly identified that token synchronization and distributed tool execution need to land atomically: otherwise ranks can diverge before reaching the tool-result collective. #291 now includes this PR's synchronized sampling implementation and tests together with rank-zero tool execution.
